### PR TITLE
bugfix: completions for implicit methods

### PR DESCRIPTION
**Pull request by kasiaMarek**
resolves: https://github.com/scalameta/metals/issues/5893

### CODE REVIEW

**Comment by filipwiech** [comment]
Hey @kasiaMarek, thank you for this, after the #5904 I really wanted to have better implicit completions in Scala 2 too. :+1:  Does this PR replace the work done by @tgodzik in #6002, or is it something related, but not exactly the same? Asking just out of curiosity, to gauge how far the support for the aging Scala 2 will improve. :wink:

File: tests/cross/src/test/scala/tests/pc/CompletionExtensionMethodSuite.scala [hunk 1]
```diff
@@ -427,41 +508,119 @@ class CompletionExtensionMethodSuite extends BaseCompletionSuite {
...
+       |    def testOps(b: Int) = ???
+       |  }
+       |}
+       |
+       |object ActualTest {
+       |  A(1).tes@@
+       |}
+       |""".stripMargin,
+    "testOps(b: Int): Nothing (implicit)",
+    filter = _.contains("testOps")
```
Comment (jkciesluk) [inline tests/cross/src/test/scala/tests/pc/CompletionExtensionMethodSuite.scala:623]: To make sure there are no incorrect completions?
```suggestion
    filter = _.contains("implicit")
```

**Review by jkciesluk — APPROVED** [review]
LGTM

File: mtags/src/main/scala-2/scala/meta/internal/pc/completions/FuzzyUpperBound.scala [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+trait FuzzyUpperBound { this: MetalsGlobal =>
+  implicit class XtensionType(tpe: Type) {
+    def fuzzy_<:<(tpeWithParams: TypeWithParams) = {
+      def adjustedType =
+        performSubstitutions(tpeWithParams.tpe, tpeWithParams.typeParams)
+
+      tpe <:< tpeWithParams.tpe || tpe <:< adjustedType
+    }
+
+    private def performSubstitutions(
```
Comment (tgodzik) [inline mtags/src/main/scala-2/scala/meta/internal/pc/completions/FuzzyUpperBound.scala:14]: Could you add some explanations here?

File: tests/cross/src/test/scala/tests/pc/CompletionSnippetSuite.scala [hunk 3]
```diff
@@ -219,7 +219,7 @@ class CompletionSnippetSuite extends BaseCompletionSuite {
       "2.13" ->
         """|Iterable
            |Iterable[$0] {}
-           |IterableOnce[$0] {}
+           |IterableOnce[$0]
```
[the remark this answers is not in the document]
Comment (kasiaMarek) [reply tests/cross/src/test/scala/tests/pc/CompletionSnippetSuite.scala:222]: That's caused by `presentationCompiler.restart()`. This test only ever worked when the whole suite was run, not just this test. I'll create an issue to investigate this.

File: mtags/src/main/scala-2/scala/meta/internal/pc/completions/FuzzyUpperBound.scala [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+          tpe match {
+            case TypeRef(pre, sym, args) =>
+              TypeRef(pre, sym, args.map(performSubstitutions(_, typeParams)))
+            case t => t
+          }
+      }
+    }
+  }
+
+  case class TypeWithParams(tpe: Type, typeParams: List[Symbol])
```
[the remark this answers is not in the document]
Comment (kasiaMarek) [reply mtags/src/main/scala-2/scala/meta/internal/pc/completions/FuzzyUpperBound.scala]: That's a good point actually. I changed this and it should much cleaner now.

**Comment by filipwiech** [comment]
> > Does this PR replace the work done by @tgodzik in #6002, or is it something related, but not exactly the same?
> 
> It replaces the changes. Is there any specific case you see lacking in this PR connect to implicit completions?

Not sure to be honest, I was thinking about "indirect" (chained) implicit completions, as well as completions for type class instances, but maybe those are either already covered by the changes or out of scope (no pun intended). :slightly_smiling_face: 

Some quick code to visualize, sorry if it's not clear enough:
```scala
trait Format[T] {

  def format(input: T): String

}

object Format {

  def apply[T](implicit ev: Format[T]): Format[T] = ev

}

object Foo {

  implicit class IntOps(private val i: Int) extends AnyVal {

    def inc: Int = i + 1
    def dec: Int = i - 1
    def print: String = s"($i)"

  }

  implicit class StringOps(private val s: String) extends AnyVal {

    def quote: String = s""""$s""""
    def howMany: Int = s.size

  }

  implicit val formatForInt: Format[Int] = new Format[Int] {

    override def format(input: Int): String = s"int = $input"

  }

  implicit def formatForOther[T]: Format[T] = new Format[T] {

    override def format(input: T): String = input.toString()

  }

  implicit class FormatOps[T](private val f: Format[T]) extends AnyVal {

    def formatBetter(input: T): String = s">>>${f.format(input)}<<<"

  }

}

object Bar {

  //import Foo.* <-- not in scope

  val a = 1.inc.dec.print.quote.howMany
  val b = Format[Int].formatBetter(1)
  val c = Format[String].formatBetter("test")

}
```

**Comment by kasiaMarek** [comment]
> I was thinking about "indirect" (chained) implicit completions

There isn't much of a difference if it's chained or not. But it will suggest the completions with an import one by one (`import Foo.IntOps`, `import Foo.StringOps`) and won't be able to add `import Foo.*`.

> completions for type class instances

That's a whole different story. If I understand this correctly, it's more of a search implicit values by type. I think it might be better to get something like this using a code action on implicit value missing error.

**Pull request merged**